### PR TITLE
perf: optimize fft batch

### PR DESCRIPTION
**Pull request by batzor**
# Description

In this PR, we try to optimize our FFTBatch algorithm since it showed a slow performance on Ubuntu for some reason.

### CODE REVIEW

File: tachyon/math/matrix/matrix_utils.h [hunk 1]
```diff
@@ -155,7 +160,14 @@ void ReverseMatrixIndexBits(Eigen::MatrixBase<Derived>& mat) {
   OMP_PARALLEL_FOR(size_t row = 1; row < rows; ++row) {
     size_t ridx = base::bits::ReverseBitsLen(row, log_n);
     if (row < ridx) {
-      mat.row(row).swap(mat.row(ridx));
+      auto row1 = absl::MakeSpan(
+          reinterpret_cast<uint8_t*>(mat.derived().data() + row * mat.cols()),
+          mat.cols() * sizeof(typename Derived::Scalar));
+      auto row2 = absl::MakeSpan(
+          reinterpret_cast<uint8_t*>(mat.derived().data() + ridx * mat.cols()),
+          mat.cols() * sizeof(typename Derived::Scalar));
```
Comment (chokobole) [inline tachyon/math/matrix/matrix_utils.h:168]:
```suggestion
      absl::Span<uint8_t> row1(
          reinterpret_cast<uint8_t*>(mat.derived().data() + row * mat.cols()),
          mat.cols() * sizeof(typename Derived::Scalar));
      absl::Span<uint8_t> row2(
          reinterpret_cast<uint8_t*>(mat.derived().data() + ridx * mat.cols()),
          mat.cols() * sizeof(typename Derived::Scalar));
```

**Comment by chokobole** [comment]
- [refac(benchmark): generate trace in FFTBatch benchmark](https://github.com/kroma-network/tachyon/pull/524/commits/303e4beb9008a2b546108ad61f4834dab2fc2963) 
- [fix(math): improve trace readability](https://github.com/kroma-network/tachyon/pull/524/commits/ddc1c2197ddfe3125dcd34dd7132c0b674bec58f) 
shouldn't they be typed as **feat**?
-[refac(benchmark): reduce memory usage](https://github.com/kroma-network/tachyon/pull/524/commits/0a52c1d01d608bc2dadbe623411950c427d51ef7)
should be typed as **perf**. Also please apply this change to fft benchmark, too.
- [fix(math): fix FFTBatch unittest](https://github.com/kroma-network/tachyon/pull/524/commits/433b67b3006c3b35e90979a2abf0c19b4ced3d13)
should be typed as **test**. But why did you change?
- [chore(benchmark): update benchmark](https://github.com/kroma-network/tachyon/pull/524/commits/86ac0b67ab33dd2a5742aae3d8dd3ceb80cd376c) should be typed as **docs**

File: tachyon/math/matrix/matrix_utils.h (hunk 1 above)
Comment (chokobole) [reply tachyon/math/matrix/matrix_utils.h:168]: Please add something like `if constexpr (Derived::Options == Eigen::RowMajor)`. I am not sure about the condition, please take a look at the code in Eigen matrix.

File: tachyon/math/polynomials/univariate/radix2_evaluation_domain_unittest.cc [hunk 2]
```diff
@@ -25,7 +25,7 @@ TYPED_TEST_SUITE(Radix2EvaluationDomainTest, PrimeFieldTypes);
 
 TYPED_TEST(Radix2EvaluationDomainTest, FFTBatch) {
   using F = TypeParam;
-  for (uint32_t log_r = 0; log_r < 5; ++log_r) {
+  for (uint32_t log_r = 1; log_r < 5; ++log_r) {
```
Comment (chokobole) [inline tachyon/math/polynomials/univariate/radix2_evaluation_domain_unittest.cc:28]: Why did you change this line?

Comment (batzor) [reply tachyon/math/polynomials/univariate/radix2_evaluation_domain_unittest.cc]: It was due to edge case in FFTBatch. I will change it and use `EXPECT_DEATH` here.

File: benchmark/fft_batch/fft_batch_benchmark.cc [hunk 3]
```diff
@@ -94,6 +95,11 @@ void Run(const FFTBatchConfig& config) {
 }
 
 int RealMain(int argc, char** argv) {
+  base::Profiler profiler(
+      {base::FilePath("/tmp/fft_batch_benchmark.perfetto-trace")});
+  profiler.Init();
+  profiler.Start();
```
[the remark this answers is not in the document]
Comment (batzor) [reply benchmark/fft_batch/fft_batch_benchmark.cc:98]: I will add to all benchmark :)

File: tachyon/math/polynomials/univariate/radix2_evaluation_domain.h [hunk 4]
```diff
@@ -98,6 +98,8 @@ class Radix2EvaluationDomain : public UnivariateEvaluationDomain<F, MaxDegree>,
     if constexpr (F::Config::kModulusBits > 32) {
       NOTREACHED();
     }
+    CHECK_GT(roots_vec_.size(), size_t{0});
+    CHECK_GT(packed_roots_vec_.size(), size_t{0});
```
Comment (dongchangYoo) [inline tachyon/math/polynomials/univariate/radix2_evaluation_domain.h:102]: These changes don't seem to be related to the title of this commit.

File: tachyon/math/polynomials/univariate/radix2_evaluation_domain_unittest.cc [hunk 5]
```diff
@@ -33,8 +33,12 @@ TYPED_TEST(Radix2EvaluationDomainTest, FFTBatch) {
...
     std::unique_ptr<Radix2EvaluationDomain<F>> domain =
         Radix2EvaluationDomain<F>::Create(size_t{1} << log_r);
-    domain->FFTBatch(result);
-    EXPECT_EQ(expected, result);
+    if (log_r == 0) {
+      EXPECT_DEATH(domain->FFTBatch(result), "");
+    } else {
+      domain->FFTBatch(result);
+      EXPECT_EQ(expected, result);
+    }
```
Comment (dongchangYoo) [inline tachyon/math/polynomials/univariate/radix2_evaluation_domain_unittest.cc:41]: This also seems to be changes that don't match the title of the commit.

File: tachyon/math/polynomials/univariate/radix2_evaluation_domain_unittest.cc [hunk 6]
```diff
@@ -49,8 +53,12 @@ TYPED_TEST(Radix2EvaluationDomainTest, CosetLDEBatch) {
...
     std::unique_ptr<Radix2EvaluationDomain<F>> domain =
         Radix2EvaluationDomain<F>::Create(size_t{1} << log_r);
-    domain->CosetLDEBatch(result, 1, shift);
-    EXPECT_EQ(expected, result);
+    if (log_r == 0) {
+      EXPECT_DEATH(domain->CosetLDEBatch(result, 1, shift), "");
+    } else {
+      domain->CosetLDEBatch(result, 1, shift);
+      EXPECT_EQ(expected, result);
+    }
```
Comment (dongchangYoo) [inline tachyon/math/polynomials/univariate/radix2_evaluation_domain_unittest.cc:61]: This also seems to be changes that don't match the contents of the commit too.

Changing the commit title might also be one way to address this.

**Review by dongchangYoo — CHANGES_REQUESTED** [review]

File: benchmark/fft_batch/plonky3/src/lib.rs [hunk 7]
```diff
@@ -8,19 +8,18 @@ use tachyon_rs::math::finite_fields::baby_bear::BabyBear as CppBabyBear;
...
-    let n = 1 << n_log;
     let size = n * batch_size;
     let values: Vec<BabyBear> = unsafe { Vec::from_raw_parts(data as *mut BabyBear, size, size) };
 
     let messages = RowMajorMatrix::<BabyBear>::new(values, batch_size);
     let dft = Radix2DitParallel::default();
 
     let start = Instant::now();
-    let mut dft_result = dft.dft_batch(messages).to_row_major_matrix();
+    let dft_result = dft.dft_batch(messages).to_row_major_matrix();
```
Comment (TomTaehoonKim) [inline benchmark/fft_batch/plonky3/src/lib.rs:22]: Can you make this change a separate commit?

File: benchmark/fft_batch/README.md
Comment (ashjeong) [inline benchmark/fft_batch/README.md:1]: Is it possible to run all of these tests on the same degrees across machines for easier comparison?

**Review by ashjeong — COMMENTED** [review]
https://github.com/kroma-network/tachyon/pull/524/commits/fe0d196592b6e08ba73edfc8d40dd9a53af7a95f

Could you put a comment or add a commit body explaining/linking to the reason for the use of -03?

File: benchmark/fft_batch/README.md
Comment (batzor) [reply benchmark/fft_batch/README.md:1]: Hmm for now if we just run low degree tests, it may seem that Tachyon is slower in production. But in fact, Tachyon is faster in a real world use case.

Comment (ashjeong) [reply benchmark/fft_batch/README.md:1]: Is it possible to just run with the same higher degree tests then?

Comment (batzor) [reply benchmark/fft_batch/README.md:1]: WARNING: On Mac M3, tests beyond degree 24 are not feasible due to memory constraints.

I added this to the `README` :)

Comment (ashjeong) [reply benchmark/fft_batch/README.md:1]: Amazing 👌

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by ashjeong — APPROVED** [review]
LGTM

File: tachyon/math/matrix/matrix_utils.h [hunk 8]
```diff
@@ -144,8 +144,16 @@ void ExpandInPlaceWithZeroPad(Eigen::MatrixBase<Derived>& mat,
...
 template <typename Derived>
+#if HAS_ATTRIBUTE(optimize)
+void __attribute__((optimize(3)))
+ReverseMatrixIndexBits(Eigen::MatrixBase<Derived>& mat) {
+#else
 void ReverseMatrixIndexBits(Eigen::MatrixBase<Derived>& mat) {
+#endif
   TRACE_EVENT("Utils", "MatrixUtils::ReverseMatrixIndexBits");
+
+  CHECK(Derived::IsRowMajor);
```
Comment (chokobole) [inline tachyon/math/matrix/matrix_utils.h:155]:
```suggestion
  static_assert(Derived::IsRowMajor);
```

File: benchmark/fft/fft_benchmark.cc [hunk 9]
```diff
@@ -82,94 +82,94 @@ void Run(const FFTConfig& config) {
       config.exponents(),
       [](uint32_t exponent) { return base::NumberToString(exponent); }));
 
+  reporter.AddVendor(Vendor::Tachyon());
+  for (Vendor vendor : config.vendors()) {
+    reporter.AddVendor(vendor);
+  }
```
Comment (chokobole) [inline benchmark/fft/fft_benchmark.cc:88]: If we see the function in`FFTRunner::Run()` and `FFTRunner::RunExternal()`, it seems not needed.

File: benchmark/fft/fft_benchmark_gpu.cc [hunk 10]
```diff
@@ -39,46 +39,48 @@ void Run(const FFTConfig& config) {
...
-  std::vector<size_t> degrees = config.GetDegrees();
-
-  std::cout << "Generating evaluation domain and random polys..." << std::endl;
-  std::vector<std::unique_ptr<Domain>> domains =
-      base::Map(degrees, [](size_t degree) { return Domain::Create(degree); });
-  std::vector<PolyOrEvals> polys = base::Map(
-      degrees, [](size_t degree) { return PolyOrEvals::Random(degree); });
-  std::cout << "Generation completed" << std::endl;
+  reporter.AddVendor(Vendor::TachyonCPU());
+  reporter.AddVendor(Vendor::TachyonGPU());
```
Comment (chokobole) [inline benchmark/fft/fft_benchmark_gpu.cc:43]: It seems not needed because you can add vendors with `should_record` in `Run()`

File: benchmark/fft_batch/fft_batch_benchmark.cc [hunk 11]
```diff
@@ -58,36 +58,33 @@ void Run(const FFTBatchConfig& config) {
...
-        return matrix;
-      });
-
   FFTBatchRunner<Domain> runner(reporter, config);
-  runner.set_domains(absl::MakeSpan(domains));
-  runner.set_inputs(absl::MakeSpan(inputs));
 
-  std::vector<math::RowMajorMatrix<F>> results;
-  runner.Run(Vendor::Tachyon(), results, config.run_coset_lde());
+  reporter.AddVendor(Vendor::Tachyon());
```
Comment (chokobole) [inline benchmark/fft_batch/fft_batch_benchmark.cc:63]: This line can be removed with a same reason above.

File: benchmark/fft/fft_benchmark.cc (hunk 9 above)
Comment (batzor) [reply benchmark/fft/fft_benchmark.cc:88]: I removed from `Run()` and `RunExternal()` because we call it multiple times for the same vendor.

**Review by chokobole — APPROVED** [review]
LGTM
